src/io/flutter/run/FlutterRunner.java

```diff
@@ -69,9 +84,97 @@ public void sessionResumed() {
+      assert path != null; // already checked
+      dasExecutionContextId = DartAnalysisServerService.getInstance().execution_createContext(path);
+    }
+    else {
+      dasExecutionContextId = null; // remote debug or can't start DAS
+    }
+
+    try {
+      if (state instanceof FlutterAppState && ((FlutterAppState)state).getMode().isDebug()) {
+        return doExecuteDartDebug(state, env, dasExecutionContextId);
+      }
+    }
+    catch (RuntimeConfigurationError e) {
+      throw new ExecutionException(e);
+    }
+
     return super.doExecute(state, env);
   }
 
+  private RunContentDescriptor doExecuteDartDebug(final @NotNull RunProfileState state,
+                                                  final @NotNull ExecutionEnvironment env,
+                                                  final @Nullable String dasExecutionContextId) throws RuntimeConfigurationError,
+                                                                                                       ExecutionException {
+    final DartSdk sdk = DartSdk.getDartSdk(env.getProject());
+    assert (sdk != null); // already checked
+
+    final RunProfile runConfiguration = env.getRunProfile();
+    if (!(runConfiguration instanceof FlutterRunConfigurationBase)) {
+      LOG.error("Unexpected run configuration: " + runConfiguration.getClass().getName());
+      return null;
+    }
+
+    if (!(state instanceof FlutterAppState)) {
+      LOG.error("Unexpected run profile state: " + state.getClass().getName());
+      return null;
+    }
+    final FlutterAppState appState = (FlutterAppState)state;
+
+    final VirtualFile contextFileOrDir;
+    VirtualFile currentWorkingDirectory;
+    final ExecutionResult executionResult;
+    final String debuggingHost;
+    final int observatoryPort;
+
+    contextFileOrDir = ((FlutterRunConfigurationBase)runConfiguration).getRunnerParameters().getDartFile();
+
+    final String cwd =
+      ((FlutterRunConfigurationBase)runConfiguration).getRunnerParameters().computeProcessWorkingDirectory(env.getProject());
+    currentWorkingDirectory = LocalFileSystem.getInstance().findFileByPath((cwd));
+
+    executionResult = appState.execute(env.getExecutor(), this);
+
+    debuggingHost = null; //TODO: should this be set?
+    observatoryPort = appState.getObservatoryPort();
+
+    FileDocumentManager.getInstance().saveAllDocuments();
+
+    final XDebuggerManager debuggerManager = XDebuggerManager.getInstance(env.getProject());
+    final XDebugSession debugSession = debuggerManager.startSession(env, new XDebugProcessStarter() {
+      @Override
+      @NotNull
+      public XDebugProcess start(@NotNull final XDebugSession session) {
+        final DartUrlResolver dartUrlResolver = getDartUrlResolver(env.getProject(), contextFileOrDir);
+        return new FlutterDebugProcess(session,
+                                       StringUtil.notNullize(debuggingHost, "localhost"),
+                                       observatoryPort,
```

Comment:
FYI, going forward, everything taking an observatory port will need to be changed to taking a service protocol URL.